### PR TITLE
Fvp dram1

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: core/arch/arm/include/mm/core_mmu.h [hunk 1]
```diff
@@ -165,22 +165,56 @@ static inline const char *teecore_memtype_name(enum teecore_memtypes type)
...
+		paddr_size_t size;
+	};
 };
 
 #define __register_memory2(_name, _type, _addr, _size, _section, _id) \
 	static const struct core_mmu_phys_mem __phys_mem_ ## _id \
 		__used __section(_section) = \
 		{ .name = _name, .type = _type, .addr = _addr, .size = _size }
 
+#if __SIZEOF_LONG__ != __SIZEOF_PADDR__
```
Comment (etienne-lms) [inline core/arch/arm/include/mm/core_mmu.h]: looks strange to have 2 macros `register_phys_mem()` and `register_phys_mem_ul()`.
Can't use something like:
```
#if __SIZEOF_LONG__ == __SIZEOF_PADDR__
#define __PHYSMEM_LOC(_addr, _size) \
			.addr = (_addr), .size = (_size),
#else
#define __PHYSMEM_LOC(_addr, _size) \
			.lo_addr = (_addr), .lo_size = (_size),
#endif

#define __register_memory2(_name, _type, _addr, _size, _section, _id) \
	static const struct core_mmu_phys_mem __phys_mem_ ## _id \
		__used __section(_section) = \
 		{ .name = _name, .type = _type, __PHYSMEM_LOC(_addr, _size) },
```

File: core/arch/arm/include/kernel/linker.h [hunk 2]
```diff
@@ -58,15 +58,15 @@ extern const struct core_mmu_phys_mem __end_phys_mem_map_section;
 extern const struct core_mmu_phys_mem __start_phys_nsec_ddr_section;
 extern const struct core_mmu_phys_mem __end_phys_nsec_ddr_section;
 
-#define VCORE_UNPG_RX_PA	((paddr_t)__vcore_unpg_rx_start)
+#define VCORE_UNPG_RX_PA	((unsigned long)__vcore_unpg_rx_start)
```
Comment (etienne-lms) [inline core/arch/arm/include/kernel/linker.h]: As these hold physical addresses/sizes, `CFG_CORE_LARGE_PHYS_ADDR=y` requires 64bit fields here. Why doesn't `paddr_t` fit?

Comment (jenswi-linaro) [reply core/arch/arm/include/kernel/linker.h]: Static initialization based on addresses of variables seems to the a bit fragile. As soon as there's a 64-bit type on a 32-bit system involved the compiler tends to complain. `unsigned long` should be a good enough compromise for all systems.

File: core/arch/arm/include/mm/core_mmu.h (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/include/mm/core_mmu.h]: We need both the normal and the _ul version. The normal is needed when supplying 64-bit addresses from a proper constant, while the _ul version must be used when initializing from variables provided by the link script.

It's not beautiful and have some sharp corners, but I can't find a way around it.

**Comment by jenswi-linaro** [comment]
Ping?

**Comment by etienne-lms** [comment]
sorry.
Why not always using the `UL` way for `register_phys_mem` ?

**Comment by jenswi-linaro** [comment]
The UL (which is short for `unsigned long` in this context) only handles 32-bit physical addresses on a 32-bit platform. In the FVP case with AArch32 that's not enough as some physical addresses are wider than that. The addresses (or constants) from the link script can always be kept in an `unsigned long` since they are originally based on pointers or virtual addresses.

File: core/arch/arm/include/kernel/linker.h (hunk 2 above)
Comment (etienne-lms) [reply core/arch/arm/include/kernel/linker.h]: ok

**Comment by etienne-lms** [comment]
Commit: "core: add register_phys_mem_ul()"
* `Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`
  I still don't get why we need to `register_phys_mem` macros. It is not clear which one one shall use over which circumstances.

For all other commits:
* `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tags applied.

To summarize you can say that `register_phys_mem_ul()` is needed whenever `register_phys_mem()` isn't accepted by the compiler, for instance when supplying `VCORE_UNPG_RX_PA` when compiling for AArch32 and `CFG_CORE_LARGE_PHYS_ADDR` == `y`).

**Comment by jforissier** [comment]
Maybe split the EMSG line to silence the checkpatch warning?
```
                       EMSG("Non-sec mem (%#" PRIxPA ":%#" PRIxPASZ
                            ") overlaps map (type %d %#" PRIxPA ":%#zx)",
                             start[n].addr, start[n].size,
                             map->type, map->pa, map->size);
```

**Comment by jenswi-linaro** [comment]
Split the `EMSG()` line

**Comment by jforissier** [comment]
Can you please rebase manually? (conflict)

**Comment by jenswi-linaro** [comment]
Rebased

**Pull request merged**